hooks/browser_integration.py

```diff
@@ -0,0 +1,137 @@
+# Copyright (c) 2017 Shotgun Software Inc.
+# 
+# CONFIDENTIAL AND PROPRIETARY
+# 
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit 
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your 
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights 
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+Hook that provides utilities used by Toolkit's browser integration.
+"""
+
+import os
+import hashlib
+import json
+import fnmatch
+import datetime
+
+import sgtk
+
+class BrowserIntegration(sgtk.Hook):
+    def filter_engine_commands(self, commands):
+        """
+        Filters out engine commands that are not desired in the web app menus.
+
+        :param list commands: The list of engine commands to filter.
+
+        :returns: A filtered list of engine commands.
+        :rtype: list
+        """
+        # Filter out any commands that didn't come from an app. This will
+        # filter out things like the "Reload and Restart" command.
+        filtered = dict()
+
+        for name, command in commands.iteritems():
+            if command["properties"].get("app") is not None:
+                filtered[name] = command
+            else:
+                sgtk.platform.current_engine().logger.debug(
+                    "Command filtered out for browser integration: %s" % command
+                )
+        return filtered
+
+    def get_cache_lookup_hash(self, entity_type, pc_descriptor):
+        """
+        Computes a unique key for a row in a cache database for the given
+        pipeline configuration descriptor and entity type.
+
+        :param str entity_type: The entity type.
+        :param pc_descriptor: A descriptor object for the pipeline configuration.
+
+        :returns: The computed lookup hash.
+        :rtype: str
+        """
+        return "%s@%s" % (pc_descriptor.get_uri(), entity_type)
+
+    def get_cache_contents_hash(self, pc_descriptor):
+        """
+        Computes an md5 hashsum for the given pipeline configuration. This
```

Comment:
I am not sure i see the value in letting the hook compute the full cache. I think the hook should do as little as possible. This gives us more flexibility to change things. Computing the cache for env files and the descriptor object itself i think we should keep internal to the system. I started writing a scaffold hook as an illustration - it ended up being bigger than i had expected - hopefully it's helpful. It's similar to what we have talked about in the past:

```python


class WebsocketsDispatch(Hook):

    def get_site_state_fields(self):
        """
        Called when a shotgun page is loaded.

        Returns a list of Shotgun fields which are needed
        to carry out the hash computation that is used to identify
        when the server state in Shotugn has changed.

        :returns: list of Shotgun fields
        """
        hash_records = []

        hash_records.append({
            "name": "LastPipelineConfigModification", # name passed to compute_site_state_hash()
            "entity_type": "PipelineConfiguration",
            "filters": [], # can use context here?
            "fields": ["updated_at"],
            "order_by": [{"field_name": "updated_at", "direction": "desc"}]
        })

        hash_records.append({
            "name": "LastSoftwareModification", # name passed to compute_site_state_hash()
            "entity_type": "Software",
            "filters": [], # can use context here?
            "fields": ["updated_at"],
            "order_by": [{"updated_at": "updated_at", "direction": "desc"}]
        })

        # note - if this is hard to do on the sg side, we can reduce
        #        it down to be a simpler format!

        return hash_records

    def compute_site_state_hash(self, sg_hash_data):
        """
        Given the data requested in the get_site_state_fields() hook methods, 
        compute a hash and return it. This hash should represent the state of
        shotgun at this point in time, taking into account all relevant fields
        which may affect the state of the registered commands that are being 
        cached. By default, this has is carved from the most recently
        updated timestamp for the software entity and the pipeline config entity,
        meaning that any change to any of the records in those tables
        would trigger a change in this hash.

        Customization: We could imagine an (expert case) where an reporting
        app creates a list of registered commands based on a list of available
        reports in Shotgun. This use case is analogous to the app launcher 
        depending on the software entity. In this case, we want to make
        sure that we include the reports entity in the cache hash, so that 
        whenever someone changes a report setting, the menu gets regenerated.

        :returns: hash string
        """

        # input for example would be
        # sg_hash_data = {"LastPipelineConfigModification": 123123123, "LastSoftwareModification": 121211212}

        return hash(sg_hash_data)

    def get_hash_fields(self, entity_type):
        """
        This is used to determine all the shotgun fields needed for 
        an entity type in order to successfully compute a hash that 
        uniquely represents its data.

        For example, we want to store one set of cached command entries
        for tasks linked to assets and one set of entries for tasks linked 
        to Shots, so we return the entity field for tasks as well as the type
        field, allowing compute_cache_hash() to produce a hash key including
        both type and link when we deal with tasks. This means that we can 
        construct a pick_environment hook which has two configs, one for 
        shot tasks and one for asset tasks.

        The cache compuation here effectively controls the granularity 
        with which cache entries are stored and invalidated. 
        """
        if entity_type == "Task":
            return ["entity", "type"]
        else:
            return ["type"]

    def compute_cache_hash(self, sg_entity_data):
        """
        Compute the hash for a shotgun object. This hash will be used to
        look up an entry in our cached commands hash for fast lookup.
        If the hash is not found in the cache, a full command generation
        will take place and the result of this will be stored against the
        hash returned by this method.
        
        The input parameter sg_entity_data contains data about the
        object for which we want to look up registered commands.
        
        Customization: Imagine you want different launch entries to appear
        based on the department a user is in. You can do this by:
            1. adding a department field to Shot, asset etc.
            2. modify get_hash_fields() to retrieve this data
            3. compute the hash in this method and include the department
               when getting the hash key for asset and shot
            4. modify pick environment so that we have two distinct
               envirionments with different apps for different departments.

        NOTE: this hash key is combined internally with the "site hash"
              reuturned from compute_site_state_hash() and the state
              of the pipeline descriptor that contains the configuration.
              (with a more complex has computed for mutable configs)
        """
        if sg_entity_data["type"] == "Task":
            # make a key based on the entity type and the linked entity type
            hash_key = "%s_%s" % (sg_entity_data["type"], sg_entity_data["entity"]["type"])
        else:
            hash_key = sg_entity_data["type"]

        return hash(hash_key)

```

Let's discuss - maybe some of this is hard to do - but i think kind of approach just exposes the parts of the system that we want the user to take control over, and we can keep the rest internal.